cpp/src/arrow/compute/exec.cc

```diff
@@ -164,13 +171,43 @@ Result<ExecBatch> ExecBatch::Make(std::vector<Datum> values) {
     }
 
     if (length != value.length()) {
-      return Status::Invalid(
-          "Arrays used to construct an ExecBatch must have equal length");
+      // all the arrays should have the same length
+      return kInvalidValues;
     }
   }
 
-  if (length == -1) {
-    length = 1;
+  return length == -1 ? 1 : length;
+}
+
+}  // namespace
+
+std::optional<int64_t> ExecBatch::InferLength(const std::vector<Datum>& values) {
+  const int64_t length = DoInferLength(values);
+  if (length < 0) {
+    return std::nullopt;
+  }
+  return {length};
+}
+
+Result<ExecBatch> ExecBatch::Make(std::vector<Datum> values, int64_t length) {
+  // Infer the length again and/or validate the given length.
+  const int64_t inferred_length = DoInferLength(values);
+  switch (inferred_length) {
+    case kEmptyInput:
+      if (length < 0) {
+        return Status::Invalid(
+            "Cannot infer ExecBatch length without at least one value");
+      }
+      break;
+
+    case kInvalidValues:
+      return Status::Invalid(
+          "Arrays used to construct an ExecBatch must have equal length");
+
+    default:
+      DCHECK(length < 0 || length == inferred_length);
```

Comment:
I wonder if we should make `length != inferred_length` an invalid status instead of a failed DCHECK.  I could go either way on this.

Reply:
Makes sense as the caller could proceed using the mistaken `length` for other things. I will add a commit changing this.

Reply:
Done.